sdk/src/logs/batch_log_processor.cc

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "opentelemetry/sdk/logs/batch_log_processor.h"
+
+#include <iostream>
+#include <vector>
+
+using opentelemetry::logs::LogRecord;
+using opentelemetry::sdk::common::AtomicUniquePtr;
+using opentelemetry::sdk::common::CircularBuffer;
+using opentelemetry::sdk::common::CircularBufferRange;
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace logs
+{
+BatchLogProcessor::BatchLogProcessor(std::unique_ptr<LogExporter> &&exporter,
+                                     const size_t max_queue_size,
+                                     const std::chrono::milliseconds schedule_delay_millis,
+                                     const size_t max_export_batch_size)
+    : exporter_(std::move(exporter)),
+      max_queue_size_(max_queue_size),
+      schedule_delay_millis_(schedule_delay_millis),
+      max_export_batch_size_(max_export_batch_size),
+      buffer_(max_queue_size_),
+      worker_thread_(&BatchLogProcessor::DoBackgroundWork, this)
+{}
+
+void BatchLogProcessor::OnReceive(std::unique_ptr<LogRecord> &&record) noexcept
+{
+  if (is_shutdown_.load() == true)
+  {
+    return;
+  }
+
+  if (buffer_.Add(record) == false)
+  {
+    return;
+  }
+
+  // If the queue gets at least half full a preemptive notification is
+  // sent to the worker thread to start a new export cycle.
+  if (buffer_.size() >= max_queue_size_ / 2)
+  {
+    // signal the worker thread
+    cv_.notify_one();
+  }
+}
+
+bool BatchLogProcessor::ForceFlush(std::chrono::microseconds timeout) noexcept
+{
```

Comment:
With 2 vars and atomic, and two while loops -- this code is overly complex and not easy to read. It appears you can rewrite the while loops with condvar `cv.wait_for` instead.

Also, in general, across all of this code - be careful about something like:
```cpp
             auto val = x.load();
             if (val)
```

Because this `load` then `compare` is not atomic. The state of what you read might have been changed by another thread. When you assign with `load` - you no longer compare the actual state, but some prior state you fetched. It's safer to use atomic compare-and-exchange or [atomic exchange](https://en.cppreference.com/w/cpp/atomic/atomic/exchange) where applicable.

Reply:
Sorry for the delay, I had to take a while to think this through - so

1. Yup, makes sense to change the `while` loops to  `cv.wait_for` with a predicate, for readability (done)

2. Another change I have made is changing all instances of `operator=` assignment for `atomic<bool>` variables to `.store(...)` for clarity (though they are actually equivalent and both atomic - see comment below). 

2. The three instances I found of non-atomic load/assign-then-compare (in the example that you listed) are in `batch_log_processor.cc`, and here's why I think they are working correctly in our case, although they're not atomic:
* Lines 110-113 in in `DoBackgroundWork()`: Although it is not atomic, there is only 1 thread calling the `DoBackgroundWork()` function at once, which is the worker thread. There won't be any more threads created, as the worker thread is only created/assigned once to `DoBackgroundWork()` in the Processor's constructor. (I have added a comment above the function to clarify this) 
* Line 169 in `Export()`: There is no other thread calling `Exporter::Export()` so the condition here should never change. 
* Lines 177-179 of `Export()`: I removed while loop around the cv notify (`force_flush_cv_.notify_one()`) altogether, because `is_force_flush_notified_` is only being set to either true/false in two places. First it's set to true by lines 177-188 in `Export()`, then the only other place where its value is being modified to `false` is in line 88, where it is _after_ being woken up by the cv's condition on it (in line 85). 

Let me know if I'm thinking about these correctly!

Reply:
Sorry for the delay, I had to take a while to think this through - so

1. Yup, makes sense to change the `while` loops to  `cv.wait_for` with a predicate, for readability (done)

2. Another change I have made is changing all instances of `operator=` assignment for `atomic<bool>` variables to `.store(...)` for clarity (though they are actually equivalent and both atomic - see comment below). 

2. The three instances I found of non-atomic load/assign-then-compare (in the example that you listed) are in `batch_log_processor.cc`, and here's why I think they are working correctly in our case, although they're not atomic:
* Lines 110-113 in in `DoBackgroundWork()`: Although it is not atomic, there is only 1 thread calling the `DoBackgroundWork()` function at once, which is the worker thread. There won't be any more threads created, as the worker thread is only created/assigned _once_ to `DoBackgroundWork()` in the Processor's constructor. (I have added a comment above the function to clarify this) 
* Line 169 in `Export()`: There is no other thread calling `Exporter::Export()` so the condition here should never change. 
* Lines 177-179 of `Export()`: I removed while loop around the cv notify (`force_flush_cv_.notify_one()`) altogether, because `is_force_flush_notified_` is only being set to either true/false in two places. First it's set to true by lines 177-188 in `Export()`, then the only other place where its value is being modified to `false` is in line 88, where it is _after_ being woken up by the cv's condition on it (in line 85). 

Let me know if I'm thinking about these correctly!

Reply:
https://github.com/open-telemetry/opentelemetry-cpp/issues/477